internal/extensions/extensions.go

```diff
@@ -0,0 +1,238 @@
+// MicroCluster-backed services.
+//
+// To distinguish between `internal` and `external` extensions, `internal` extensions
+// are prefixed with a `internal:` value.
+// e.g, `internal:runtime_extension_v1`, `internal:runtime_extension_v2`, ...
+//
+// External extensions are related to capabilities of MicroCluster-backed services.
+// e.g, `microovn_custom_encapsulation_ip`
+type Extensions []string
+
+// Populate internal extensions here.
+var internalExtensions = &Extensions{
+	"internal:runtime_extension_v1",
+}
+
+// validateExternalExtension validates the given external extension.
+func validateExternalExtension(extension string) error {
+	if extension == "" {
+		return fmt.Errorf("Extension cannot be empty")
+	}
+
+	// Aside from the internal extensions, we let the extension to be like
+	// `<extension>` with only lowercase letters and underscores.
+	if !externalExtensionRegex.MatchString(extension) {
+		return fmt.Errorf("External extension name %q is invalid: Extension name must contain only lowercase letters and underscores, and must not begin or end with an underscore", extension)
+	}
+
+	return nil
+}
+
+// validateInternalExtension validates the given internal extension.
+func validateInternalExtension(extension string) error {
+	if extension == "" {
+		return fmt.Errorf("Extension cannot be empty")
+	}
+
+	// Internal extensions should be in the format `internal:<extension>` with only lowercase letters and underscores.
+	if !internalExtensionRegex.MatchString(extension) {
+		return fmt.Errorf("Internal extension name %q is invalid: Extension name must contain only lowercase letters and underscores, and must begin by `internal:`", extension)
+	}
+
+	return nil
+}
+
+// HasExtension reports whether the extension set supports the given extension.
+func (e *Extensions) HasExtension(ext string) bool {
+	for _, extension := range *e {
+		if extension == ext {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Version returns the number of extensions in the set, representing its version number.
+func (e *Extensions) Version() int {
+	return len(*e)
+}
+
+// NewExtensionRegistry creates a new instance of the Extensions struct, which represents a registry of extensions.
+// We have the option to populate the internal extensions or not. Having s option becomes handy
+// when we want to create a custom registry of extensions (see `NewExtensionRegistryFromList`)
+func NewExtensionRegistry(populateInternal bool) (*Extensions, error) {
+	extensions := new(Extensions)
+	var err error
+	if populateInternal {
+		for _, internalExtension := range *internalExtensions {
+			err = validateInternalExtension(internalExtension)
+			if err != nil {
+				return nil, err
+			}
+
+			i := sort.Search(len(*extensions), func(i int) bool {
+				return (*extensions)[i] >= internalExtension
+			})
+
+			*extensions = append(*extensions, internalExtension)
+			copy((*extensions)[i+1:], (*extensions)[i:])
+			(*extensions)[i] = internalExtension
```

Comment:
Why does this need to be so complex? 

Why can't we just do
```go
if populateInternal {
  # Instantiate extensions with a size of 0 but a capacity equal to the size of the internal extension list.
  extensions = make(Extensions, 0, len(*internalExtensions))
  
  for _, internalExtension := range *internalExtensions {
      extensions = append(extensions, internalExtension)
  }
}

return extensions
```

Reply:
I guess, what I wanted to do here, is to insert the extension in the registry at the right index so that these are always sorted (sorted internal extensions and then sorted external extensions):
`internal:a, internal:b, .., internal_e, not_internal_a, not_internal_b`

If we just do:

```go
for _, internalExtension := range *internalExtensions {
      // validate first
      // ... 
      extensions = append(extensions, internalExtension)
}
```
if `*internalExtensions` is not sorted, this approach will keep `extensions` not sorted. The idea of inserting the entry in a sorted way, is to keep the data predictable in the database and when comparing two extensions (instead of always calling the sort function on the registry before inserting it to the DB or when comparing the entries with an other extensions)

Reply:
Sorry, I'm not following you about the sorting. You're sorting alphabetically by name of the extension, but if we're sorting at all, I would think it should be by recency of the API extension. Is there a benefit to sorting alphanumerically?

* `internalExtensions` is a list of arbitrary strings that begin with `internal:`, the list should be ordered by when the api extension was introduced by default, as we should always be appending to the end of the list when we add a new one.

* `extensions` is being instantiated in this function so it's always empty, and by the end it should contain all the same items as `internalExtensions`.

Then you run `sort.Search` and compare `(*extensions)[i] >= internalExtension` but this would compare the actual length of the strings, but this breaks the order that the API extensions were introduced in, and instead sorts them alphanumerically. 

I'm not sure if it's such a good idea to lose that order, and it's not clear to me what we gain with alphanumeric sorting instead.

Reply:
You're right, after thinking about it, it does not really bring anything to sort it alphanumerically. I'm removing this.